Batch editing

toban opened the pull request.

toban committed:
Add batch editing

- Adds base class EditEndpoint
- Adds endpoint /batch-edit
- Adds some docs describing the payload

toban commented:
- [x] savedOtherItems might reference not needed
- [x] No exception add break
- [x] Add result only if status is good or ok
- [ ] change user to private on EditEndpoint
- [ ] persistItem can set the item revision instead of returning saved items.
- [x] just return an array on parseBatchRequests - remove BatchEditRequest
- [x] Add PHP integration test for batch-edit endpoint

toban commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -52,69 +53,31 @@ public function __construct(
-		}
-
-		// And make the edit
+	public function persistItem( ReconciledItem $reconciledItem, array $otherItems ): array {
```
Add PHPDOC

lucaswerkmeister replied:
Done

lucaswerkmeister commented:
> just return an array on parseBatchRequests - remove BatchEditRequest

done, but all of the tests still passed even though I hadn’t updated them, which means

- [x] add tests for parseBatchRequests

lucaswerkmeister commented:
on the other points:

> change user to private on EditEndpoint

not convinced this one is needed tbh. I’d be willing to merge without this

> persistItem can set the item revision instead of returning saved items.

to be done in a follow-up

lucaswerkmeister committed:
Update EditRequestParserTest test method names

The tested method was renamed from parseRequestInterface() to
parseRequestBody() in commit b813c8d51a.

guergana commented:
LGTM. Also nice now we new tests.

toban commented on includes/MediaWiki/Api/BatchEditEndpoint.php:
```diff
@@ -0,0 +1,125 @@
+				];
+			}, $status->getValue() );
+		}
+		// TODO attach errors
```
We could do this now maybe or write this down as a ticket? I suspect there might be a series of followups planned already?

lucaswerkmeister replied:
Ticket makes sense, I think. (The same comment applies to `SingleEditEndpoint`, though we don’t have a TODO there.)

toban replied:
https://phabricator.wikimedia.org/T284134

toban commented on tests/phpunit/integration/MediaWiki/Api/BatchEditEndpointTest.php:
```diff
@@ -0,0 +1,263 @@
+ * @group Database
+ * @license GPL-2.0-or-later
+ */
+class BatchEditEndpointTest extends \MediaWikiIntegrationTestCase {
```
This reminds me a lot of EditEndpoint test, not sure there are any real benefits of doing so but maybe we could refactor these afterwards to share some of this code / token tests etc if we feel there is enough time.

toban commented on tests/phpunit/unit/MediaWiki/Request/EditRequestParserTest.php:
```diff
@@ -97,7 +97,7 @@ private function getPropertyDataTypeLookup(): PropertyDataTypeLookup {
 	}
 
-	public function testParseRequestInterface_good(): void {
+	public function testParseRequestBody_good(): void {
```
good catch

toban reviewed (commented):
Can't approve this but I do! it looks great. Just some comments for possible followups.
